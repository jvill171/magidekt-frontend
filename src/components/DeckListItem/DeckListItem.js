import { useContext, useEffect } from "react";
import UserContext from "../../context/UserContext";
import { NavLink } from "react-router-dom";
import "./DeckListItem.scss"
import { identityToMana } from "../../helpers/toMana";

const DeckListItem = ({data, showOwner = true}) =>{
    const { user } = useContext(UserContext)

    const { colorIdentity,
            deckName,
            description,
            displayName:deckOwner,
            format,
            id,
            tags,
            cardCount } = data;

    // Capitalize first letter
    const upFormat = format[0].toUpperCase() + format.slice(1);
    // Generate mana icons
    const colorIcons = identityToMana(colorIdentity)

    return(
        <NavLink to={`/decks/${user}/${id}`} className="DeckListItem decklist-item">

            <p>{`/decks/${user}/${id}`}</p>

            <h3 className="deck-name">{deckName}</h3>
            <p className="deck-identity">
                {cardCount > 0 
                    ? colorIcons
                    : <span className="ms no-mana"></span>}
            </p>
            <ul>
                {showOwner &&
                    <li>
                        <b>Owner:</b> {deckOwner}
                    </li>
                }
                <li>
                    <b>Deck ID:</b> {id}
                </li>
                <li>
                    <b>Cards:</b> {cardCount}
                </li>
                <li>
                    <b>Format:</b> {upFormat}
                </li>
                <li>
                    <b>Tags:</b> {tags && tags.length > 0
                        ? tags.join(', ')
                        : ''}
                </li>
                {description && 
                    <li>
                        <b>Desc:</b> {description}
                    </li>
                }
            </ul>
        </NavLink>
    )
}

export default DeckListItem;