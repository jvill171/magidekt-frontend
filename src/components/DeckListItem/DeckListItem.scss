
.decklist-item {
    flex: 0 0 calc(25% - 10px);
    min-width: 200px; 
    background-color: rgb(240, 240, 240); 
    padding: 10px; 
    border: 1px solid rgb(99, 99, 99);
    border-radius: 5px;
    box-sizing: border-box;
    color: black;
    text-decoration: none;
    
    ul{
        text-align: left;
        list-style: none;
        margin: 10px auto;
        padding: 5px;
    }

    .deck-name{
        font-weight: bold;
        text-decoration: underline;
    }

    .deck-identity{
        font-size: 1.4em;
        margin: -10px auto;
        .ms{
            border-radius: 50%;      /* Rounded border */
            padding: calc(2em / 15); /* 2/15 ths of parent elment font size*/
            margin: 0 calc(1em/15 + 2px);     /* 1/15 ths of parent elment font size*/
            box-shadow: calc(-1em/15) calc(1em/15) 0 rgba(0,0,0,0.85);
            border: 1px solid black;
            color:black;
        }
        .ms-g{
            background-color: #abddbd;
        }
        .ms-u{
            background-color: #bae7fb;
        }
        .ms-r{
            background-color: #faba9f;
        }
        .ms-w{
            background-color: #fefddf;
        }
        .ms-b{
            background-color: #d7d0cd;
        }
        .ms-c{
            background-color: #d7d0cd;
        }
        .no-mana{
            color: transparent;
            padding: .9rem;
        }
    }
    
    &:hover{
        background-color: hsl(43, 74%, 90%); 
        scale: 103%;
    }
}
  

.description{
    width: 100%;
    p{
        height: 3em;
        padding: 5px;
        margin: 2px auto;
        border: 1px solid lightgray;
        border-radius: 5px;
        overflow-y: scroll;
    }
}