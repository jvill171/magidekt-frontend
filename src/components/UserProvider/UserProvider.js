import React, { useEffect, useState } from "react";
import jwtDecode from 'jwt-decode';
import { useNavigate } from "react-router-dom";
import UserContext from "../../context/UserContext";
import Navbar from "../Navbar/Navbar";
import Loading from "../Loading/Loading";
import RouteList from "../RouteList/RouteList";

const UserProvider = () =>{

    const [user, setUser] = useState(undefined);
    const [isLoading, setIsLoading] = useState(true)
    const navigate = useNavigate();

    const updateUser = () => {
      try{
        const token = localStorage.getItem(`_token`)
        const { username } = jwtDecode(token)
        setUser(username)
      }catch(err){
        // Authentication to routes that require authentication will be handled by backend
        setUser(undefined) // Bad/no token = Invalid user; Logged out
        navigate("/") //After logout, redirect to "/"
      }
    };
    
    useEffect(()=>{
      if(!user){
        updateUser()
      }
      setIsLoading(false)
    },[user])

    return(
        <UserContext.Provider value={{ user, updateUser }}>
        {isLoading ?
          <>
            <h1>Loading...</h1>
            <Loading />
          </>
          :
          <>
            <Navbar />
            <RouteList />
          </>
        }
        </UserContext.Provider>
    )
}

export default UserProvider;