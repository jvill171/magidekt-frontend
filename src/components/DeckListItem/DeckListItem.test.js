import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import UserProvider from '../UserProvider/UserProvider';
import DeckListItem from './DeckListItem';

//Smoke test
test('renders DeckListItem component without crashing', () => {
    render(
        <MemoryRouter>
        <UserProvider>
            <DeckListItem />
        </UserProvider>
        </MemoryRouter>
    );
});

//Snapshot test
test('DeckListItem component matches snapshot', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <UserProvider>
          <DeckListItem />
        </UserProvider>
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
});