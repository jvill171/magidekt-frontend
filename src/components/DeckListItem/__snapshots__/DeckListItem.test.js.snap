// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeckListItem component matches snapshot 1`] = `
<DocumentFragment>
  <nav
    class="navbar-fade"
  >
    <div
      class="navbar"
    >
      <div
        class="logo"
      >
        <a
          href="/"
        >
          <span>
            Magi
          </span>
          dekt
        </a>
      </div>
      <ul
        class="nav-NavLinks"
      >
        <li>
          <a
            class=""
            href="/login"
          >
            Login
          </a>
        </li>
        <li>
          <a
            class=""
            href="/signup"
          >
            Signup
          </a>
        </li>
      </ul>
    </div>
  </nav>
  <div
    class="Home"
  >
    <h1>
      Magidekt
    </h1>
    <div
      class="auth-buttons"
    >
      <a
        class=""
        href="/login"
      >
        <button>
          Login
        </button>
      </a>
      <a
        class=""
        href="/signup"
      >
        <button>
          Sign Up
        </button>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
